### ISSUE TITLE
CertLogic not JsonLogic compliant

### ISSUE THREAD

**Issue by ubamrein** [ehn-dcc-development/dgc-business-rules#67]
According to the JsonLogic-Specification, any operator takes as an inner node an array containing all parameters for the operator. Iff the operator is a unary operator, there exists the "syntactic sugar" to supply the argument directly (c.f. https://jsonlogic.com/operations.html#var).

The current CertLogic implementation unnecessarily changes JsonLogic-Syntax, which makes it diverse from all implementations based on JsonLogic (e.g. the swift implementation https://github.com/eu-digital-green-certificates/dgc-certlogic-ios)

Further, the specification says: `As noted before: the var data access operation is the only type of expression that can have a non-array argument specification` (c.f. [here](https://github.com/ehn-dcc-development/dgc-business-rules/blob/main/certlogic/specification/README.md#data-access-var)), where `can` should be equivalent to [may](https://datatracker.ietf.org/doc/html/rfc2119), which means it should be optional to have `non-array` arguments.

In the `CertLogic.kt` class on the other hand the `non-array` argument is enforced:
https://github.com/ehn-dcc-development/dgc-business-rules/blob/16fa11db7ea7b02b754c1911770eedda45e91c03/certlogic/certlogic-kotlin/src/main/kotlin/eu/ehn/dcc/certlogic/certlogic.kt#L8

**Comment by dslmeinte** [comment]
That's intentional: having the syntax sugar-version makes sense because data access is so fundamental to how CertLogic is used. It makes less sense to have both the sugared and unsugared versions at the same time: offering multiple versions is usually just confusing to users.

I understand the draw of consistency and symmetry here. The purpose of CertLogic is to specify a language that's as small as possible, while still being adequate to express business rules for DCCs. The addition of the date-time comparison operations, as well as the `extractFromUVCI` operation, makes CertLogic not strictly a subset of JsonLogic.

CertLogic is only compatible with JsonLogic on the syntax they both support. Outside of that there's a big difference in behaviour, especially because CertLogic tends to throw `Error`s on constructs that JsonLogic supports (although not necessarily consistently across JsonLogic implementations).

**Comment by ubamrein** [comment]
Well, but it is never mentioned. So if it is intentional, then be specific about it in the documents. It is already hard enough following two organizations, multiple repositories, a billion documents and a trillion discussions. The only "divergence" ever mentioned are the operators added (c.f. [here](https://github.com/ehn-dcc-development/dgc-business-rules/blob/main/certlogic/specification/README.md)).

Why use JsonLogic, and always mention JsonLogic, if you just can't use a JsonLogic-backend for CertLogic?

Also not offering the _non_ syntactic sugar in the CertLogic, but saying ` Implementors of the DCC validator using a JsonLogic implementation instead of a CertLogic implementation need to provide these custom operations to JsonLogic as well - see the first paragraph of this document.` will definitely lead to more problems, as implementors strictly using a JsonLogic-backend, could be confused, as to why their rules lead to parse errors on other platforms.

In my opinion, either support JsonLogic as a language (with non syntactic-sugar as well) or just don't mention JsonLogic, and don't use it anywhere. 

> CertLogic is only compatible with JsonLogic on the syntax they both support. Outside of that there's a big difference in behaviour, especially because CertLogic tends to throw Errors on constructs that JsonLogic supports (although not necessarily consistently across JsonLogic implementations).

This basically also reads: Don't use JsonLogic, as I can always say "Well language `A` and language `B` are only compatible on the syntax they both support". 

Always talking about JsonLogic, but not providing support for the most basic language feature just leads to more confusion.

TL;DR;
Remove all mentions of JsonLogic in the documentation (or at least provide an up-to-date list of what exactly is different), or just add support for array parameters (which is probably less effort)

**Comment by dslmeinte** [comment]
The reason to not use JsonLogic is that the various implementations (for different platforms, in different languages) already diverge (quite a bit). It's also not possible to use one JsonLogic implementation (e.g. the JavaScript one), because it's not always possible to run a JS engine inside an app.

Adding support for `{ "var": [...] }` might seem minor, but it does expand the language, so also the specification, the test suite, and all currently-available implementations (JS/TS, Kotlin, Dart). I like to keep the language really as small as possible. It's always possible to "desugar" a "sugared" version of the language to CertLogic as it is now.

But I can improve the specification in the way you suggest: keep references to JsonLogic to a minimum, and be explicit about differences (including the `{ "var": [...] }` difference).

**Comment by ubamrein** [comment]
> The reason to not use JsonLogic is that the various implementations (for different platforms, in different languages) already diverge (quite a bit).

Can you specify how they diverge, and how this diversion is different to the various implementation for CertLogic (and/or the "validated" json-logic libraries found [here](https://github.com/ehn-dcc-development/dgc-business-rules/blob/main/documentation/implementations.md)). I just find it hard to understand, how implementing a language all over again is _easier_ to maintain than dealing with possible differences. 

The [documentation](https://github.com/ehn-dcc-development/dgc-business-rules/blob/main/certlogic/specification/README.md#data-access-var) should also be fixed, as the statement should not be 

> As noted before: the var data access operation is the only type of expression that can have a non-array argument specification.

but rather 

> Note: the var data access operation , contraire to other operators, _must_ be used with a _single_ string argument

Also, going on with unary operators: Why does "!" not allow the syntactic-sugar? Following your argument of creating an easy and concise subset, "!" should also only take a single argument. Currently, the argument _must_ be an array, and it takes the first element (which also contradicts the specification [here](https://github.com/ehn-dcc-development/dgc-business-rules/blob/main/certlogic/specification/README.md#negation-), which states that it takes `one operand`, which in the context of "CertLogic" and the `var` simplification reads as "one argument" and not an array, whereof the first element is interpreted as an operand).

My bottom line is, why invent a language, which is not human readable, based on something existing, but then not using that existing implementation, and change the syntax in non documented ways? This is what I would call "[...] is usually just confusing to users".

Also if the final point here is that CertLogic is the ground truth, then the iOS swift reference implementation has to be adjusted, as otherwise there is inconsistent behaviour across platforms (which is also confusing to the users _and_ developers), and all references to "json-logic implementation that pass the test-suite" should be removed.

**Comment by dslmeinte** [comment]
The syntax of the operations available in JsonLogic is _not_ changed in CertLogic: not all syntax may be supported, but that's not “change” as such.

Implementing a language such as CertLogic (especially if it's as small as it is) takes very little effort compared to aligning the implementations of an existing, but poorly-defined language. Anecdotal evidence: JsonLogic has a test suite that covers exactly _none_ of the issues I immediately ran into when trying to use it for business rules. E.g., the Kotlin implementation has a problem with parsing a string value that happens to be interpretable as a double value.

To be able to provide a language to define business/validation rules on top of the DCC within a very short time frame, making a new, minimal language with some compatibility with JsonLogic made more sense than trying to align JsonLogic implementations, and then still needing to extend all of those with custom operations.

And consider what would (have to) happen when countries would use the “saltier” parts of JsonLogic, where it's even easier to venture outside the boundaries of the standard test suite? The burden of interoperability lies with the language implementors, not with the business rules authors.

I'm happy to make clarifying modifications to the CertLogic specification and documentation around it to address this issue.

**Comment by ubamrein** [comment]
> The syntax of the operations available in JsonLogic is not changed in CertLogic: not all syntax may be supported, but that's not “change” as such.

Forcing one specific syntactic-sugar on one operator but not on the other sounds to me like a syntax change...

> Implementing a language such as CertLogic (especially if it's as small as it is) takes very little effort compared to aligning the implementations of an existing, but poorly-defined language

I'm not sure if I follow here. Compared to what I find on CertLogic, [JsonLogic](https://jsonlogic.com/) seems to be more well-defined, consistent and better documented... But that might just be my view on the things...

> And consider what would (have to) happen when countries would use the “saltier” parts of JsonLogic [...]

I'm not saying CertLogic should implement every edge case coming up in JsonLogic. I just beg for _basic_ syntax support. For example: I still don't understand the reasoning (and you did not even provide any arguments for or against it) why `var` obviously and to avoid confusion, needs to only support the syntactic sugar, where as `!` is not allowed to use the syntactic sugar, although both of them are used exactly in the same setting, and the implementation could have been exactly the same (e.g. dropping support for default values in the `var` case), by just using the first argument if it is an array.

I was just sticking to JsonLogic as it has a much nicer documentation including examples for all operators, whereas I could not find something similar on the EU site (which makes it hard to understand, see the example above for "one operand"). But be it, then I will make an issue on the swift-cert-logic repo, in order to align the implementations for Android and iOS...

**Comment by dslmeinte** [comment]
As a language engineer, I can tell you that JsonLogic is definitely not more well-defined, consistent, and better documented than CertLogic: this is exactly why I started CertLogic in the first place. JsonLogic is fine if you stick to one implementation, and you are using it just to implement business logic (which you supposedly test in some way), but it's woefully inadequate if you want to make such logic interchangeable (as is our use case). Then you really need a tight spec, a validator, etc. CertLogic has that, JsonLogic not.  Don't let “nice”-looking documentation (thanks to some CSS) fool you.

Nevertheless, I appreciate your contributions.

**Comment by ubamrein** [comment]
I mean it is not the "nice"-looking documentation, in fact I actually find it ugly :D It's rather the table of examples after each operator, which helps a lot. 

Maybe I can extend the documentation with similar examples :)

**Comment by dslmeinte** [comment]
I would appreciate it mightily if you could extend the documentation (either the spec itself, or a more human-readable “Getting started”-type document) with examples!
